cpp/include/rapidsmpf/buffer/resource.hpp

```diff
@@ -478,4 +518,32 @@ MemoryReservation reserve_or_fail(
     std::optional<MemoryType> const& preferred_mem_type = std::nullopt
 );
 
+/**
+ * @brief Acquire a memory reservation and execute a function, which will ensure
+ * the reservation is released when the function goes out of scope.
+ *
+ * Similar to `BufferResource::with_reservation`, but this function would not reserve
+ * memory, but holds on to the reservation until the function is executed. Is useful in
+ * situations where the memory reservation is made with spilling.
+ *
+ * @param reservation moved memory reservation.
+ * @param f The function to execute.
+ *
+ * @return The result of the function.
+ */
+auto with_memory_reservation(MemoryReservation&& reservation, auto&& f) {
+    using F = std::decay_t<decltype(f)>;
+
+    if constexpr (std::invocable<F, MemoryReservation&>) {
+        return std::forward<F>(f)(reservation);
+    } else if constexpr (std::invocable<F>) {
+        return std::forward<F>(f)();
+    } else {
+        static_assert(
+            [] { return false; }(),
```

Comment:
```suggestion
            std::integral_constant<F, false>(),
```

Your example happens to work today but a sufficiently smart compiler might decide to compile-time evaluate the lambda and then you run afoul of the issue that `static_assert(false)` is not allowed. See also https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2593r0.html